File: scripts/species_images/upload_species_images_to_supabase.py
import json
import mimetypes
import os
import time
import urllib.error
import urllib.request
from pathlib import Path


ROOT = Path(__file__).resolve().parents[2]
IMAGE_DIR = ROOT / "database" / "seed" / "species_images"
INPUT_PATH = IMAGE_DIR / "final_species_image_assets.json"
OUTPUT_PATH = IMAGE_DIR / "final_species_image_assets_uploaded.json"
PROGRESS_PATH = IMAGE_DIR / "python_upload_progress.json"
REPORT_PATH = IMAGE_DIR / "python_upload_report.json"
BUCKET = "species-images"
DEFAULT_DELAY_MS = 250


def read_json(path, fallback):
    if not path.exists():
        return fallback
    with path.open("r", encoding="utf-8") as handle:
        return json.load(handle)


def write_json(path, value):
    temp_path = path.with_suffix(path.suffix + ".tmp")
    with temp_path.open("w", encoding="utf-8") as handle:
        json.dump(value, handle, indent=2)
        handle.write("\n")
    temp_path.replace(path)


def get_env_config():
    supabase_url = os.environ.get("SUPABASE_URL") or os.environ.get("EXPO_PUBLIC_SUPABASE_URL")
    service_role_key = os.environ.get("SUPABASE_SERVICE_ROLE_KEY")

    if not supabase_url or not service_role_key:
        raise RuntimeError(
            "Missing Supabase config. Set SUPABASE_URL or EXPO_PUBLIC_SUPABASE_URL, plus SUPABASE_SERVICE_ROLE_KEY."
        )

    return {
        "supabase_url": supabase_url.rstrip("/"),
        "service_role_key": service_role_key,
    }


def parse_positive_int(value, fallback=None):
    if value is None or value == "":
        return fallback
    try:
        parsed = int(value)
    except ValueError:
        return fallback
    return parsed if parsed > 0 else fallback


def upload_delay_seconds():
    delay_ms = parse_positive_int(os.environ.get("SPECIES_UPLOAD_DELAY_MS"), DEFAULT_DELAY_MS)
    return delay_ms / 1000


def upload_limit():
    return parse_positive_int(os.environ.get("SPECIES_UPLOAD_LIMIT"))


def request_json(url, method, config, body=None):
    headers = {
        "Authorization": f"Bearer {config['service_role_key']}",
        "apikey": config["service_role_key"],
    }
    data = None
    if body is not None:
        data = json.dumps(body).encode("utf-8")
        headers["Content-Type"] = "application/json"

    request = urllib.request.Request(url, data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(request, timeout=60) as response:
            response.read()
            return {"ok": True, "status": response.status, "body": ""}
    except urllib.error.HTTPError as error:
        body_text = error.read().decode("utf-8", errors="replace")
        return {"ok": False, "status": error.code, "body": body_text[:500]}
    except urllib.error.URLError as error:
        return {"ok": False, "status": None, "body": str(error.reason)}


def ensure_bucket(config):
    bucket_url = f"{config['supabase_url']}/storage/v1/bucket/{BUCKET}"
    check = request_json(bucket_url, "GET", config)
    if check["ok"]:
        return {"bucket_status": "exists"}

    bucket_not_found = check["status"] == 404 or "Bucket not found" in check["body"]
    if not bucket_not_found:
        return {
            "bucket_status": "check_failed",
            "bucket_error": f"HTTP {check['status']}: {check['body']}",
        }

    create_url = f"{config['supabase_url']}/storage/v1/bucket"
    create = request_json(
        create_url,
        "POST",
        config,
        {"id": BUCKET, "name": BUCKET, "public": True},
    )
    if create["ok"]:
        return {"bucket_status": "created"}

    return {
        "bucket_status": "create_failed",
        "bucket_error": f"HTTP {create['status']}: {create['body']}",
    }


def progress_key(asset):
    return asset.get("species_key") or asset.get("storage_path")


def load_assets():
    source_assets = read_json(INPUT_PATH, [])
    uploaded_assets = read_json(OUTPUT_PATH, [])
    progress_rows = read_json(PROGRESS_PATH, [])
    uploaded_by_key = {}

    for row in uploaded_assets + progress_rows:
        if row.get("public_url"):
            uploaded_by_key[progress_key(row)] = row

    merged_assets = []
    for asset in source_assets:
        merged_assets.append({**asset, **uploaded_by_key.get(progress_key(asset), {})})

    return source_assets, merged_assets


def save_uploaded_manifest(source_assets, uploaded_by_key):
    merged_assets = []
    for asset in source_assets:
        merged_assets.append({**asset, **uploaded_by_key.get(progress_key(asset), {})})

    progress_rows = sorted(uploaded_by_key.values(), key=lambda row: str(progress_key(row)))
    write_json(OUTPUT_PATH, merged_assets)
    write_json(PROGRESS_PATH, progress_rows)


def object_path_for(asset):
    species_key = asset.get("species_key", "")
    asset_status = asset.get("asset_status")
    local_path = asset.get("local_file_path") or ""
    extension = Path(local_path).suffix.lower() or ".jpg"

    if asset_status == "ready_ai_placeholder":
        return f"ai/{species_key}.png"
    if asset_status == "ready_real_image":
        return f"real/{species_key}{extension}"

    return asset.get("storage_path") or f"ai/{species_key}.png"


def content_type_for(file_path):
    guessed, _ = mimetypes.guess_type(str(file_path))
    if guessed:
        return guessed
    if file_path.suffix.lower() == ".png":
        return "image/png"
    return "image/jpeg"


def public_url_for(config, object_path):
    return f"{config['supabase_url']}/storage/v1/object/public/{BUCKET}/{object_path}"


def upload_file(config, asset, object_path, file_path):
    upload_url = f"{config['supabase_url']}/storage/v1/object/{BUCKET}/{object_path}"
    headers = {
        "Authorization": f"Bearer {config['service_role_key']}",
        "apikey": config["service_role_key"],
        "Content-Type": content_type_for(file_path),
        "x-upsert": "true",
    }

    with file_path.open("rb") as handle:
        body = handle.read()

    request = urllib.request.Request(upload_url, data=body, headers=headers, method="POST")
    try:
        with urllib.request.urlopen(request, timeout=120) as response:
            response.read()
            return {"ok": True, "status": response.status, "body": ""}
    except urllib.error.HTTPError as error:
        body_text = error.read().decode("utf-8", errors="replace")
        return {"ok": False, "status": error.code, "body": body_text[:500]}
    except urllib.error.URLError as error:
        return {"ok": False, "status": None, "body": str(error.reason)}


def failure_row(asset, object_path, message):
    return {
        "species_key": asset.get("species_key"),
        "storage_path": object_path,
        "upload_status": "failed",
        "upload_error": message,
    }


def build_report(started_at, finished_at, bucket_result, total_rows, uploaded_this_run, skipped, failures, public_url_count):
    failed_keys = [row["species_key"] for row in failures]
    return {
        "startedAt": started_at,
        "finishedAt": finished_at,
        "bucket": BUCKET,
        **bucket_result,
        "totalRows": total_rows,
        "uploadedThisRun": uploaded_this_run,
        "skippedAlreadyUploaded": skipped,
        "failedThisRun": len(failures),
        "publicUrlCount": public_url_count,
        "pendingCount": total_rows - public_url_count,
        "failedSpeciesKeys": failed_keys,
        "failures": failures,
    }


def main():
    config = get_env_config()
    delay_seconds = upload_delay_seconds()
    limit = upload_limit()
    started_at = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
    source_assets, current_assets = load_assets()
    uploaded_by_key = {
        progress_key(row): row
        for row in current_assets
        if row.get("public_url")
    }
    failures = []
    uploaded_this_run = 0
    skipped = 0
    attempted = 0

    bucket_result = ensure_bucket(config)
    if bucket_result["bucket_status"].endswith("failed"):
        report = build_report(
            started_at,
            time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            bucket_result,
            len(source_assets),
            0,
            0,
            [
                failure_row(asset, object_path_for(asset), "Bucket unavailable; upload not attempted.")
                for asset in current_assets
                if not asset.get("public_url")
            ],
            len(uploaded_by_key),
        )
        write_json(REPORT_PATH, report)
        print(json.dumps(report, indent=2))
        raise SystemExit(1)

    print(
        f"Uploading {len(source_assets)} species image assets sequentially "
        f"with {int(delay_seconds * 1000)}ms delay."
    )
    if limit:
        print(f"SPECIES_UPLOAD_LIMIT active: at most {limit} new upload attempts.")

    for index, asset in enumerate(current_assets, start=1):
        key = progress_key(asset)
        if asset.get("public_url") or uploaded_by_key.get(key, {}).get("public_url"):
            skipped += 1
            print(f"[{index}/{len(current_assets)}] skipped already uploaded {asset.get('species_key')}")
            continue

        if limit and attempted >= limit:
            break

        attempted += 1
        object_path = object_path_for(asset)
        local_file_path = asset.get("local_file_path") or ""
        file_path = ROOT / local_file_path

        if not file_path.exists() or file_path.stat().st_size == 0:
            failures.append(failure_row(asset, object_path, f"Missing or empty local file: {local_file_path}"))
            print(f"[{index}/{len(current_assets)}] failed {asset.get('species_key')}")
            time.sleep(delay_seconds)
            continue

        result = upload_file(config, asset, object_path, file_path)
        if result["ok"]:
            uploaded_asset = {
                **asset,
                "storage_path": object_path,
                "storage_bucket": BUCKET,
                "public_url": public_url_for(config, object_path),
                "upload_status": "uploaded",
                "uploaded_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            }
            uploaded_by_key[key] = uploaded_asset
            uploaded_this_run += 1
            save_uploaded_manifest(source_assets, uploaded_by_key)
            print(f"[{index}/{len(current_assets)}] uploaded {asset.get('species_key')}")
        else:
            failures.append(
                failure_row(
                    asset,
                    object_path,
                    f"HTTP {result['status']}: {result['body']}",
                )
            )
            print(f"[{index}/{len(current_assets)}] failed {asset.get('species_key')}")

        time.sleep(delay_seconds)

    save_uploaded_manifest(source_assets, uploaded_by_key)
    public_url_count = len(uploaded_by_key)
    report = build_report(
        started_at,
        time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        bucket_result,
        len(source_assets),
        uploaded_this_run,
        skipped,
        failures,
        public_url_count,
    )
    write_json(REPORT_PATH, report)
    print(json.dumps(report, indent=2))

    if failures or public_url_count != len(source_assets):
        raise SystemExit(1)


if __name__ == "__main__":
    main()

File: scripts/species_images/upload_species_images_to_supabase.js
const fs = require('fs');
const path = require('path');

const ROOT = path.resolve(__dirname, '..', '..');
const IMAGE_DIR = path.join(ROOT, 'database', 'seed', 'species_images');
const INPUT_PATH = path.join(IMAGE_DIR, 'final_species_image_assets.json');
const OUTPUT_PATH = path.join(IMAGE_DIR, 'final_species_image_assets_uploaded.json');
const REPORT_PATH = path.join(IMAGE_DIR, 'upload_species_images_report.json');
const PROGRESS_PATH = path.join(IMAGE_DIR, 'upload_progress.json');
const FAILED_ROWS_PATH = path.join(IMAGE_DIR, 'upload_failed_rows.json');
const BUCKET = process.env.SPECIES_IMAGES_BUCKET || 'species-images';
const DEFAULT_CONCURRENCY = 1;
const DEFAULT_DELAY_MS = 200;
const WINDOWS_MAX_CONCURRENCY = 2;

function parsePositiveInteger(value, fallback) {
  const parsed = Number.parseInt(String(value || ''), 10);
  return Number.isFinite(parsed) && parsed > 0 ? parsed : fallback;
}

function parseCliOption(name) {
  const prefix = `--${name}=`;
  const arg = process.argv.slice(2).find((item) => item.startsWith(prefix));
  return arg ? arg.slice(prefix.length) : undefined;
}

function getUploadOptions() {
  const requestedConcurrency = parsePositiveInteger(
    parseCliOption('concurrency') || process.env.SPECIES_UPLOAD_CONCURRENCY,
    DEFAULT_CONCURRENCY,
  );
  const concurrency =
    process.platform === 'win32'
      ? Math.min(requestedConcurrency, WINDOWS_MAX_CONCURRENCY)
      : requestedConcurrency;
  const delayMs = parsePositiveInteger(
    parseCliOption('delay-ms') || process.env.SPECIES_UPLOAD_DELAY_MS,
    DEFAULT_DELAY_MS,
  );

  return {
    concurrency,
    delayMs,
    requestedConcurrency,
    concurrencyCapped: concurrency !== requestedConcurrency,
  };
}

function readJsonFile(filePath, fallback) {
  if (!fs.existsSync(filePath)) return fallback;
  try {
    return JSON.parse(fs.readFileSync(filePath, 'utf8'));
  } catch (error) {
    throw new Error(`Could not read ${path.relative(ROOT, filePath)}: ${error.message}`);
  }
}

function writeJsonFile(filePath, value) {
  const tempPath = `${filePath}.tmp`;
  fs.writeFileSync(tempPath, JSON.stringify(value, null, 2) + '\n');
  fs.renameSync(tempPath, filePath);
}

function delay(ms) {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}

function contentTypeForFile(filePath) {
  const extension = path.extname(filePath).toLowerCase();
  if (extension === '.png') return 'image/png';
  if (extension === '.webp') return 'image/webp';
  return 'image/jpeg';
}

function getSupabaseConfig() {
  const supabaseUrl = process.env.EXPO_PUBLIC_SUPABASE_URL || process.env.SUPABASE_URL;
  const serviceRoleKey = process.env.SUPABASE_SERVICE_ROLE_KEY;

  if (!supabaseUrl || !serviceRoleKey) {
    throw new Error(
      'Missing Supabase config. Set EXPO_PUBLIC_SUPABASE_URL or SUPABASE_URL, plus SUPABASE_SERVICE_ROLE_KEY.',
    );
  }

  return { supabaseUrl: supabaseUrl.replace(/\/$/, ''), serviceRoleKey };
}

async function readResponseBody(response) {
  const body = await response.text();
  return body.length > 500 ? `${body.slice(0, 500)}...` : body;
}

async function ensureBucket(config) {
  const bucketUrl = `${config.supabaseUrl}/storage/v1/bucket/${BUCKET}`;
  const getResponse = await fetch(bucketUrl, {
    headers: {
      Authorization: `Bearer ${config.serviceRoleKey}`,
      apikey: config.serviceRoleKey,
    },
  });

  if (getResponse.ok) {
    return { bucket_status: 'exists' };
  }

  if (getResponse.status !== 404) {
    return {
      bucket_status: 'check_failed',
      bucket_error: `HTTP ${getResponse.status}: ${await readResponseBody(getResponse)}`,
    };
  }

  const createResponse = await fetch(`${config.supabaseUrl}/storage/v1/bucket`, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${config.serviceRoleKey}`,
      apikey: config.serviceRoleKey,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      id: BUCKET,
      name: BUCKET,
      public: true,
      file_size_limit: null,
    }),
  });

  if (createResponse.ok) {
    return { bucket_status: 'created' };
  }

  return {
    bucket_status: 'create_failed',
    bucket_error: `HTTP ${createResponse.status}: ${await readResponseBody(createResponse)}`,
  };
}

function publicUrlFor(config, storagePath) {
  return `${config.supabaseUrl}/storage/v1/object/public/${BUCKET}/${storagePath}`;
}

async function uploadAsset(config, asset) {
  const absolutePath = path.join(ROOT, asset.local_file_path || '');
  if (!asset.local_file_path || !fs.existsSync(absolutePath)) {
    return {
      ...asset,
      upload_status: 'failed',
      upload_error: 'Missing local image file.',
    };
  }

  const storagePath = asset.storage_path;
  const uploadUrl = `${config.supabaseUrl}/storage/v1/object/${BUCKET}/${storagePath}`;
  let fileBuffer = fs.readFileSync(absolutePath);

  try {
    const response = await fetch(uploadUrl, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${config.serviceRoleKey}`,
        apikey: config.serviceRoleKey,
        'Content-Type': contentTypeForFile(absolutePath),
        'x-upsert': 'true',
      },
      body: fileBuffer,
    });

    if (!response.ok) {
      const body = await readResponseBody(response);
      return { ...asset, upload_status: 'failed', upload_error: `HTTP ${response.status}: ${body}` };
    }

    return {
      ...asset,
      upload_status: 'uploaded',
      storage_bucket: BUCKET,
      public_url: publicUrlFor(config, storagePath),
      uploaded_at: new Date().toISOString(),
    };
  } finally {
    fileBuffer = null;
  }
}

function getProgressKey(asset) {
  return asset.species_key || asset.storage_path;
}

function mergeUploadedRows(assets, progressMap) {
  return assets.map((asset) => {
    const progress = progressMap.get(getProgressKey(asset));
    return progress ? { ...asset, ...progress } : asset;
  });
}

function loadProgress(assets) {
  const progressRows = readJsonFile(PROGRESS_PATH, []);
  const uploadedRows = readJsonFile(OUTPUT_PATH, []);
  const progressMap = new Map();

  for (const row of [...uploadedRows, ...progressRows]) {
    if (row && row.public_url) {
      progressMap.set(getProgressKey(row), row);
    }
  }

  return {
    progressMap,
    uploadedAssets: mergeUploadedRows(assets, progressMap),
  };
}

function saveProgress(assets, progressMap) {
  const progressRows = [...progressMap.values()].sort((a, b) =>
    String(getProgressKey(a)).localeCompare(String(getProgressKey(b))),
  );
  writeJsonFile(PROGRESS_PATH, progressRows);
  writeJsonFile(OUTPUT_PATH, mergeUploadedRows(assets, progressMap));
}

async function runWorker({ workerId, assets, config, options, progressMap, failures, counters }) {
  while (counters.nextIndex < assets.length) {
    const index = counters.nextIndex;
    counters.nextIndex += 1;
    const asset = assets[index];
    const progressKey = getProgressKey(asset);
    const progress = progressMap.get(progressKey);

    if (progress && progress.public_url) {
      counters.skippedAlreadyUploaded += 1;
      console.log(`[${index + 1}/${assets.length}] skipped already uploaded ${asset.species_key}`);
      continue;
    }

    try {
      const result = await uploadAsset(config, asset);
      if (result.upload_status === 'uploaded' && result.public_url) {
        progressMap.set(progressKey, result);
        counters.uploadedThisRun += 1;
        saveProgress(assets, progressMap);
        console.log(`[${index + 1}/${assets.length}] uploaded ${asset.species_key}`);
      } else {
        failures.push(result);
        counters.failedThisRun += 1;
        console.log(`[${index + 1}/${assets.length}] failed ${asset.species_key}`);
      }
    } catch (error) {
      const failedRow = {
        ...asset,
        upload_status: 'failed',
        upload_error: error instanceof Error ? error.message : String(error),
      };
      failures.push(failedRow);
      counters.failedThisRun += 1;
      console.log(`[${index + 1}/${assets.length}] failed ${asset.species_key}`);
    }

    if (options.delayMs > 0 && counters.nextIndex < assets.length) {
      await delay(options.delayMs + workerId * 25);
    }
  }
}

function buildReport({ startedAt, bucketResult, assets, progressMap, failures, options }) {
  const uploadedAssets = mergeUploadedRows(assets, progressMap);
  const publicUrlCount = uploadedAssets.filter((asset) => asset.public_url).length;
  const failedKeys = new Set(failures.map((asset) => getProgressKey(asset)));
  const pendingRows = uploadedAssets.filter((asset) => !asset.public_url && !failedKeys.has(getProgressKey(asset)));

  return {
    startedAt,
    finishedAt: new Date().toISOString(),
    bucket: BUCKET,
    ...bucketResult,
    concurrency: options.concurrency,
    requestedConcurrency: options.requestedConcurrency,
    concurrencyCapped: options.concurrencyCapped,
    delayMs: options.delayMs,
    totalAssets: assets.length,
    successfulUploads: publicUrlCount,
    uploadedThisRun: options.counters.uploadedThisRun,
    skippedAlreadyUploaded: options.counters.skippedAlreadyUploaded,
    failedUploads: failures.length,
    pendingUploads: pendingRows.length,
    publicUrlCount,
    failures: failures.map((asset) => ({
      species_key: asset.species_key,
      storage_path: asset.storage_path,
      upload_status: asset.upload_status,
      upload_error: asset.upload_error,
    })),
    pending: pendingRows.map((asset) => ({
      species_key: asset.species_key,
      storage_path: asset.storage_path,
    })),
  };
}

async function main() {
  const config = getSupabaseConfig();
  const options = getUploadOptions();
  const assets = JSON.parse(fs.readFileSync(INPUT_PATH, 'utf8'));
  const startedAt = new Date().toISOString();
  const { progressMap } = loadProgress(assets);
  const failures = [];
  const counters = {
    nextIndex: 0,
    uploadedThisRun: 0,
    skippedAlreadyUploaded: 0,
    failedThisRun: 0,
  };
  options.counters = counters;

  const bucketResult = await ensureBucket(config);
  if (bucketResult.bucket_status.endsWith('failed')) {
    const report = {
      startedAt,
      finishedAt: new Date().toISOString(),
      bucket: BUCKET,
      ...bucketResult,
      concurrency: options.concurrency,
      delayMs: options.delayMs,
      totalAssets: assets.length,
      successfulUploads: progressMap.size,
      uploadedThisRun: 0,
      skippedAlreadyUploaded: 0,
      failedUploads: assets.length - progressMap.size,
      pendingUploads: assets.length - progressMap.size,
      publicUrlCount: progressMap.size,
      failures: assets
        .filter((asset) => !progressMap.get(getProgressKey(asset))?.public_url)
        .map((asset) => ({
          species_key: asset.species_key,
          storage_path: asset.storage_path,
          upload_status: 'not_attempted_bucket_unavailable',
        })),
    };
    writeJsonFile(REPORT_PATH, report);
    writeJsonFile(FAILED_ROWS_PATH, report.failures);
    console.log(JSON.stringify(report, null, 2));
    process.exit(1);
  }

  if (options.concurrencyCapped) {
    console.log(
      `Requested concurrency ${options.requestedConcurrency} capped to ${options.concurrency} on Windows.`,
    );
  }
  console.log(
    `Uploading ${assets.length} species image assets with concurrency ${options.concurrency} and ${options.delayMs}ms delay.`,
  );

  const workers = Array.from({ length: options.concurrency }, (_, workerIndex) =>
    runWorker({
      workerId: workerIndex,
      assets,
      config,
      options,
      progressMap,
      failures,
      counters,
    }),
  );
  await Promise.all(workers);

  saveProgress(assets, progressMap);
  if (failures.length) {
    writeJsonFile(FAILED_ROWS_PATH, failures);
  } else {
    writeJsonFile(FAILED_ROWS_PATH, []);
  }

  const report = buildReport({
    startedAt,
    bucketResult,
    assets,
    progressMap,
    failures,
    options,
  });
  writeJsonFile(REPORT_PATH, report);
  console.log(JSON.stringify(report, null, 2));

  if (report.failedUploads || report.pendingUploads || report.publicUrlCount !== assets.length) {
    process.exit(1);
  }
}

main().catch((error) => {
  console.error(error.message);
  process.exit(1);
});
